src/metadata/statistics.rs

```diff
@@ -0,0 +1,334 @@
+            null_count,
+            histogram,
+        }
+    }
+
+    pub fn column_index(&self) -> ColumnIndex {
+        self.col_idx
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn min(&self) -> Datum {
+        self.min
+    }
+
+    pub fn max(&self) -> Datum {
+        self.max
+    }
+
+    pub fn null_count(&self) -> f64 {
+        self.null_count
+    }
+
+    pub fn histogram(&self) -> &Option<Histogram> {
+        &self.histogram
+    }
+}
+
+impl Metadata for ColumnStats {}
+
+#[derive(Clone, Debug)]
+pub struct Statistics {
+    output_row_count: f64,
+
+    /// Statistics of columns, column index -> column stat
+    column_stats: Vec<Box<dyn Metadata>>,
+}
+
+impl Statistics {
+    pub const fn new(output_row_count: f64, column_stats: Vec<Box<dyn Metadata>>) -> Self {
+        Self {
+            output_row_count,
+            column_stats,
+        }
+    }
+
+    pub fn column_stats(&self) -> &Vec<Box<dyn Metadata>> {
+        &self.column_stats
+    }
+}
+
+impl Stats for Statistics {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn should_update(&self, new_stats: &Rc<dyn Stats>) -> bool {
+        let new_stats = new_stats.as_any().downcast_ref::<Statistics>().unwrap();
+        new_stats.output_row_count < self.output_row_count
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RelationStats {
+    name: String,
+    rows: f64,
+    empty: bool,
+
+    col_stat_mdids: Vec<Box<dyn MdId>>,
+}
+
+impl RelationStats {
+    pub const fn new(name: String, rows: f64, empty: bool, col_stat_mdids: Vec<Box<dyn MdId>>) -> Self {
+        Self {
+            name,
+            rows,
+            empty,
+            col_stat_mdids,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn rows(&self) -> f64 {
+        self.rows
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.empty
+    }
+
+    pub fn col_stat_mdids(&self) -> &[Box<dyn MdId>] {
+        &self.col_stat_mdids
+    }
+}
+
+impl Metadata for RelationStats {}
+
+#[derive(Clone, Debug)]
+pub struct ColumnMetadata {
+    name: String,
+    attno: u64,
+    nullable: bool,
+    width: u32,
+    default: Datum,
+}
+
+impl ColumnMetadata {
+    pub const fn new(name: String, attno: u64, nullable: bool, width: u32, default: Datum) -> Self {
+        Self {
+            name,
+            attno,
+            nullable,
+            width,
+            default,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn attno(&self) -> u64 {
+        self.attno
+    }
+
+    pub fn nullable(&self) -> bool {
+        self.nullable
+    }
+
+    pub fn width(&self) -> u32 {
+        self.width
+    }
+
+    pub fn default(&self) -> Datum {
+        self.default
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum StorageType {
+    Heap,
+    External,
+    Virtual,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum DistributionPolicyType {
```

Comment:
StorageType and DistributionPolicyType can be removed in order to simplify the business, we can consider them later.